File: adt2unit/pas/init_songdata.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure init_songdata;

{$IFNDEF ADT2PLAY}
var
  temp: Byte;
{$ENDIF} // NOT DEFINED(ADT2PLAY)

begin
  _dbg_enter ({$I %FILE%}, 'init_songdata');

  If (play_status <> isStopped) then
    begin
{$IFNDEF ADT2PLAY}
      fade_out_playback(FALSE);
{$ENDIF} // NOT DEFINED(ADT2PLAY)
      stop_playing;
    end
  else init_buffers;

  FillChar(songdata,SizeOf(songdata),0);
  FillChar(songdata.pattern_order,SizeOf(songdata.pattern_order),$080);
{$IFDEF ADT2PLAY}
  FillChar(pattdata^,PATTERN_SIZE*max_patterns,0);
{$ELSE} // NOT DEFINED(ADT2PLAY)
  For temp := 1 to max_patterns DIV 8 do
    FillChar(pattdata^[PRED(temp)],8*PATTERN_SIZE,0);
{$ENDIF} // NOT DEFINED(ADT2PLAY)

{$IFDEF ADT2PLAY}
  songdata.patt_len := 64;
  songdata.nm_tracks := 9;
{$ELSE} // NOT DEFINED(ADT2PLAY)
  songdata.patt_len := patt_len;
  songdata.nm_tracks := nm_tracks;
{$ENDIF} // NOT DEFINED(ADT2PLAY)
  songdata.tempo := tempo;
  songdata.speed := speed;
{$IFDEF ADT2PLAY}
  songdata.macro_speedup := 1;
{$ELSE} // NOT DEFINED(ADT2PLAY)
  songdata.macro_speedup := init_macro_speedup;
{$ENDIF} // NOT DEFINED(ADT2PLAY)
  speed_update := FALSE;
{$IFNDEF ADT2PLAY}
  IRQ_freq_shift := 0;
  playback_speed_shift := 0;
  songdata.bpm_data.rows_per_beat := mark_line;
{$ENDIF} // NOT DEFINED(ADT2PLAY)
  songdata.bpm_data.tempo_finetune := IRQ_freq_shift;
  lockvol := FALSE;
  panlock := FALSE;
  lockVP  := FALSE;
  tremolo_depth := 0;
  vibrato_depth := 0;
  volume_scaling := FALSE;

{$IFNDEF ADT2PLAY}
  old_chan_pos := 1;
  old_hpos := 1;
  old_page := 0;
  old_block_chan_pos := 1;
  old_block_patt_hpos := 1;
  old_block_patt_page := 0;
  marking := FALSE;
{$ENDIF} // NOT DEFINED(ADT2PLAY)

{$IFDEF ADT2PLAY}
  If (songdata.nm_tracks <= 18) then
{$ELSE} // NOT DEFINED(ADT2PLAY)
  If (nm_tracks <= 18) then
{$ENDIF} // NOT DEFINED(ADT2PLAY)
    begin
      percussion_mode := FALSE;
      _chan_n := _chmm_n;
      _chan_m := _chmm_m;
      _chan_c := _chmm_c;
    end
  else
    begin
      percussion_mode := TRUE;
      _chan_n := _chpm_n;
      _chan_m := _chpm_m;
      _chan_c := _chpm_c;
    end;

{$IFNDEF ADT2PLAY}
  For temp := 1 to 255 do
    songdata.instr_names[temp] := ' iNS_'+byte2hex(temp)+#247' ';

  For temp := 0 to $7f do
    songdata.pattern_names[temp] := ' PAT_'+byte2hex(temp)+'  '#247' ';

  songdata_crc_ord := Update32(songdata.pattern_order,
                               SizeOf(songdata.pattern_order),0);
  module_archived := TRUE;
{$ENDIF} // NOT DEFINED(ADT2PLAY)

  _dbg_leave; //EXIT //init_songdata
end;

File: adt2unit/init_songdata.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2026 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

void init_songdata (tFIXED_SONGDATA *song)
{
  DBG_ENTER ("init_songdata");

  if (play_status != isStopped)
  {
#if !ADT2PLAY
    fade_out_playback (false);
#endif // !ADT2PLAY
    stop_playing ();
  }
  else
    init_buffers ();

  memset (song, 0, sizeof (*song));
  memset (&song->pattern_order, PATTERN_ORDER_JUMP + 0, sizeof (song->pattern_order));
  memset (pattdata, 0, PATTERN_SIZE * max_patterns);

#if ADT2PLAY
  song->patt_len = 64;
  song->nm_tracks = 9;
#else // !ADT2PLAY
  song->patt_len = patt_len;
  song->nm_tracks = nm_tracks;
#endif // !ADT2PLAY
  song->tempo = tempo;
  song->speed = speed;
#if ADT2PLAY
  song->macro_speedup = 1;
#else // !ADT2PLAY
  song->macro_speedup = init_macro_speedup;
#endif // !ADT2PLAY
  speed_update = false;
#if !ADT2PLAY
  IRQ_freq_shift = 0;
  playback_speed_shift = 0;
  song->bpm_data.rows_per_beat = mark_line;
#endif // !ADT2PLAY
  song->bpm_data.tempo_finetune = IRQ_freq_shift;
  lockvol = false;
  panlock = false;
  lockVP  = false;
  tremolo_depth = 0;
  vibrato_depth = 0;
  volume_scaling = false;

#if !ADT2PLAY
  old_chan_pos = 1;
  old_hpos = 1;
  old_page = 0;
  old_block_chan_pos = 1;
  old_block_patt_hpos = 1;
  old_block_patt_page = 0;
  marking = false;
#endif // !ADT2PLAY

  if (song->nm_tracks <= MELODIC_CHANNELS_MAX)
  {
    percussion_mode = false;
    memcpy (_chan_n, _chmm_n, sizeof (_chan_n));
    memcpy (_chan_m, _chmm_m, sizeof (_chan_m));
    memcpy (_chan_c, _chmm_c, sizeof (_chan_c));
  }
  else
  {
    percussion_mode = true;
    memcpy (_chan_n, _chpm_n, sizeof (_chan_n));
    memcpy (_chan_m, _chpm_m, sizeof (_chan_m));
    memcpy (_chan_c, _chpm_c, sizeof (_chan_c));
  }

#if !ADT2PLAY
  for (unsigned i = 0; i < INSTRUMENTS_MAX; i++)
  {
    SetLength (song->instr_names[i], 9);
    snprintf (GetStr ((char *)song->instr_names[i]), 9+1,
              " iNS_%02"PRIX8"%c ", i + 1, /*charmap.*/'\xF7');
  }

  for (unsigned i = 0; i < PATTERNS_MAX; i++)
  {
    SetLength (song->pattern_names[i], 11);
    snprintf (GetStr ((char *)song->pattern_names[i]), 11+1,
              " PAT_%02"PRIX8"  %c ", i + 1, /*charmap.*/'\xF7');
  }

  songdata_crc_ord = Update32 (song->pattern_order,
                               sizeof (song->pattern_order), /*CRC32_INITVAL*/ 0);
  module_archived = true;
#endif // !ADT2PLAY

  DBG_LEAVE (); //EXIT //init_songdata
}
